### DISCUSSION TITLE
Allowing private fields initialization in constructor with System.Text.Json

### DISCUSSION THREAD

**Discussion by mickturn** [dotnet/runtime#55036] [Ideas] [answered]
Hi,

Take a look [here](https://github.com/dotnet/runtime/issues/54854) and [here](https://github.com/dotnet/runtime/discussions/54984).

Seems that @vintzl failed to explain what we want.

I have added a repro [here](https://github.com/mickturn/deserialization).

Ultimately, we just want to know what you can do to get private fields populated in the constructor like stated in the repro.

Thank you for your attention.

**Comment by vintzl** [comment]
You wast your time.

You will just have useless and poor answers from eiriktsarpalis.

**Comment by jeffhandley** [comment]
Hello,

I have been following along on the other threads, wherein @vintzl has been quite disrespectful to @eiriktsarpalis. I have hidden the comment that @vintzl made here too, as it was disruptive.

While the repro added here captures the scenario described in a standalone manner, it doesn't change the answer or guidance that can be provided. For that reason, I am going to lock this conversation and reply back on https://github.com/dotnet/runtime/discussions/54984.

-Jeff

**Comment by terrajobst** [comment, marked as answer]
@mickturn 

Quoting your document here, for easier reading:

> # Private fields deserialization
> 
> Json objects look like this:
> 
> `
> {
>     "id" : 1,
>     "name" : "test"
> }
> `
> 
> Here, the private fields are being populated in the constructor allowing complete object deserialization.
> 
> But it works only when we add accessors that are not themselves serialized... Without them you get:
> > 'Each parameter in constructor 'Void .ctor(UInt16, System.String)' on type 'Experiments.Test' must bind to an object property or field on deserialization. Each parameter name must match with a property or field on the object. The match can be case-insensitive.'
> 
> Why not allowing private fields initialization in constructor without the need of these pretty useless accessors?

You can make it work but you need to have public properties. Serialization, by its nature, is part of your public API. Serialization needs to be robust. Code like this seems very suspicious to me:

```C#
public class Test
{
    private readonly ushort _id;
    private readonly string _name;

    public Test(ushort id, string name)
    {
        _id = id;
        _name = name;
    }
}
```

Did the author rally expect to have those fields filled in deserialization? It seems very odd to allow fields to be deserialized that can't be serialized. If you don't like to see the `Id` and `Name` properties in IntelliSense, just hide them:

```C#
public class Test
{
    public Test(ushort id, string name)
    {
        Id = id;
        Name = name;
    }

    [EditorBrowsable(EditorBrowsableState.Never)]
    public ushort Id { get; }

    [EditorBrowsable(EditorBrowsableState.Never)]
    public string Name { get; }
}
```
